<a href="https://colab.research.google.com/github/Thomas-George-T/Coursera_Capstone/blob/master/Tale_of_Two_Cities_A_modern_Take.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1 align="center"> A Tale of Two cities - A Modern Take </h1>

@Author: Thomas George Thomas

In [ ]:
import pandas as pd
import requests
import numpy as np
import matplotlib.cm as cm
import matplotlib.colors as colors
import folium

# import k-means for the clustering stage
from sklearn.cluster import KMeans

# London

## STEP 1: Get the Neighbourhoods of London

### Data Collection

To get the neighbourhoods in london, we start by scraping the list of areas of london wiki page.

In [ ]:
url_london = "https://en.wikipedia.org/wiki/List_of_areas_of_London"
wiki_london_url = requests.get(url_london)
wiki_london_url

<Response [200]>

Response 200 means that we are able to make the connection

In [ ]:
wiki_london_data = pd.read_html(wiki_london_url.text)
wiki_london_data

[                                                   0
 0  Map all coordinates in "Category:Areas of Lond...
 1                 Download coordinates as: KML · GPX,
             Location                     London borough  ... Dial code OS grid ref
 0         Abbey Wood              Bexley, Greenwich [7]  ...       020    TQ465785
 1              Acton  Ealing, Hammersmith and Fulham[8]  ...       020    TQ205805
 2          Addington                         Croydon[8]  ...       020    TQ375645
 3         Addiscombe                         Croydon[8]  ...       020    TQ345665
 4        Albany Park                             Bexley  ...       020    TQ478728
 ..               ...                                ...  ...       ...         ...
 528         Woolwich                          Greenwich  ...       020    TQ435795
 529   Worcester Park       Sutton, Kingston upon Thames  ...       020    TQ225655
 530  Wormwood Scrubs             Hammersmith and Fulham  ...       020    TQ2258

Scraping the webpage gives us all the tables present on the page. We need the 2nd table, so selecting the 2nd table.

In [ ]:
wiki_london_data = wiki_london_data[1]
wiki_london_data

Location                     London borough  ... Dial code OS grid ref
0         Abbey Wood              Bexley, Greenwich [7]  ...       020    TQ465785
1              Acton  Ealing, Hammersmith and Fulham[8]  ...       020    TQ205805
2          Addington                         Croydon[8]  ...       020    TQ375645
3         Addiscombe                         Croydon[8]  ...       020    TQ345665
4        Albany Park                             Bexley  ...       020    TQ478728
..               ...                                ...  ...       ...         ...
528         Woolwich                          Greenwich  ...       020    TQ435795
529   Worcester Park       Sutton, Kingston upon Thames  ...       020    TQ225655
530  Wormwood Scrubs             Hammersmith and Fulham  ...       020    TQ225815
531          Yeading                         Hillingdon  ...       020    TQ115825
532         Yiewsley                         Hillingdon  ...       020    TQ063804

[533 rows x 6 columns]

### Data Preprocessing

we remove the spaces in the column titles and then we add _ between words.

In [ ]:
wiki_london_data.rename(columns=lambda x: x.strip().replace(" ", "_"), inplace=True)
wiki_london_data

Location                     London borough  ... Dial code OS_grid_ref
0         Abbey Wood              Bexley, Greenwich [7]  ...       020    TQ465785
1              Acton  Ealing, Hammersmith and Fulham[8]  ...       020    TQ205805
2          Addington                         Croydon[8]  ...       020    TQ375645
3         Addiscombe                         Croydon[8]  ...       020    TQ345665
4        Albany Park                             Bexley  ...       020    TQ478728
..               ...                                ...  ...       ...         ...
528         Woolwich                          Greenwich  ...       020    TQ435795
529   Worcester Park       Sutton, Kingston upon Thames  ...       020    TQ225655
530  Wormwood Scrubs             Hammersmith and Fulham  ...       020    TQ225815
531          Yeading                         Hillingdon  ...       020    TQ115825
532         Yiewsley                         Hillingdon  ...       020    TQ063804

[533 rows x 6 columns]

We see that few columns have no '_' between the words despite applying our function meaning that there are special characters

### Feature Selection

We need only the boroughs, Postal codes, Post town for further steps. We can drop the locations, dial codes and OS grid.

In [ ]:
df1 = wiki_london_data.drop( [ wiki_london_data.columns[0], wiki_london_data.columns[4], wiki_london_data.columns[5] ], axis=1)

In [ ]:
df1.head()

London borough       Post_town Postcode district
0              Bexley, Greenwich [7]          LONDON               SE2
1  Ealing, Hammersmith and Fulham[8]          LONDON            W3, W4
2                         Croydon[8]         CROYDON               CR0
3                         Croydon[8]         CROYDON               CR0
4                             Bexley  BEXLEY, SIDCUP         DA5, DA14

let's rename the Postcode district column and the london borough to something simpler

In [ ]:
df1.columns = ['borough','town','post_code']
df1

borough            town  post_code
0                Bexley, Greenwich [7]          LONDON        SE2
1    Ealing, Hammersmith and Fulham[8]          LONDON     W3, W4
2                           Croydon[8]         CROYDON        CR0
3                           Croydon[8]         CROYDON        CR0
4                               Bexley  BEXLEY, SIDCUP  DA5, DA14
..                                 ...             ...        ...
528                          Greenwich          LONDON       SE18
529       Sutton, Kingston upon Thames  WORCESTER PARK        KT4
530             Hammersmith and Fulham          LONDON        W12
531                         Hillingdon           HAYES        UB4
532                         Hillingdon    WEST DRAYTON        UB7

[533 rows x 3 columns]

Let's remove the Square brackets [ ] and numbers from the borough column

In [ ]:
df1['borough'] = df1['borough'].map(lambda x: x.rstrip(']').rstrip('0123456789').rstrip('['))
df1

borough            town  post_code
0                Bexley, Greenwich           LONDON        SE2
1    Ealing, Hammersmith and Fulham          LONDON     W3, W4
2                           Croydon         CROYDON        CR0
3                           Croydon         CROYDON        CR0
4                            Bexley  BEXLEY, SIDCUP  DA5, DA14
..                              ...             ...        ...
528                       Greenwich          LONDON       SE18
529    Sutton, Kingston upon Thames  WORCESTER PARK        KT4
530          Hammersmith and Fulham          LONDON        W12
531                      Hillingdon           HAYES        UB4
532                      Hillingdon    WEST DRAYTON        UB7

[533 rows x 3 columns]

Take the dimension of the dataframe

In [ ]:
df1.shape

(533, 3)

We currently have 533 records and 3 columns of our data. It's time to perform Feature Engineering

### Feature Engineering

We can only focusing on the neighbourhoods of London, so performing the changes

In [ ]:
df1 = df1[df1['town'].str.contains('LONDON')]
df1

borough                    town post_code
0                Bexley, Greenwich                   LONDON       SE2
1    Ealing, Hammersmith and Fulham                  LONDON    W3, W4
6                              City                  LONDON       EC3
7                       Westminster                  LONDON       WC2
9                           Bromley                  LONDON      SE20
..                              ...                     ...       ...
523                       Redbridge                  LONDON  IG8, E18
524       Redbridge, Waltham Forest  LONDON, WOODFORD GREEN       IG8
527                          Barnet                  LONDON       N12
528                       Greenwich                  LONDON      SE18
530          Hammersmith and Fulham                  LONDON       W12

[310 rows x 3 columns]

In [ ]:
df1.shape

(310, 3)

We now have only 310 rows. We can proceed with our further steps. Getting some descriptive statistics

In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 310 entries, 0 to 530
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   borough    310 non-null    object
 1   town       310 non-null    object
 2   post_code  310 non-null    object
dtypes: object(3)
memory usage: 9.7+ KB


### STEP 2: Get Geolocations of the London Neighbourhoods

We need to get the geographical co-ordinates for the neighbourhoods to plot out map. We will use the arcgis package to do so. 

Arcgis doesn't have a limitation on the number of API calls made so it fits our use case perfectly.

In [ ]:
pip install arcgis

  Using cached https://files.pythonhosted.org/packages/c5/8f/51e89ce52a085483359217bc72cdbf6e75ee595d5b1d4b5ade40c7e018b8/jsonschema-3.2.0-py2.py3-none-any.whl
  Found existing installation: jsonschema 2.6.0
    Uninstalling jsonschema-2.6.0:
      Successfully uninstalled jsonschema-2.6.0


In [ ]:
from arcgis.geocoding import geocode
from arcgis.gis import GIS
gis = GIS()

Defining London arcgis geocode function to return latitude and longitude

In [ ]:
def get_x_y_uk(address1):
   lat_coords = 0
   lng_coords = 0
   g = geocode(address='{}, London, England, GBR'.format(address1))[0]
   lng_coords = g['location']['x']
   lat_coords = g['location']['y']
   return str(lat_coords) +","+ str(lng_coords)

Checking sample data

In [ ]:
c = get_x_y_uk('SE2')

In [ ]:
c

'51.492450000000076,0.12127000000003818'

Looks good, We Copy over the postal codes of london to pass it into the geolocator function that we just defined above

In [ ]:
geo_coordinates_uk = df1['post_code']    
geo_coordinates_uk

0           SE2
1        W3, W4
6           EC3
7           WC2
9          SE20
         ...   
523    IG8, E18
524         IG8
527         N12
528        SE18
530         W12
Name: post_code, Length: 310, dtype: object

Passing postal codes of london to get the geographical co-ordinates

In [ ]:
coordinates_latlng_uk = geo_coordinates_uk.apply(lambda x: get_x_y_uk(x))
coordinates_latlng_uk

0       51.492450000000076,0.12127000000003818
1        51.51324000000005,-0.2674599999999714
6       51.51200000000006,-0.08057999999994081
7       51.51651000000004,-0.11967999999995982
9       51.41009000000008,-0.05682999999993399
                        ...                   
523    51.589770000000044,0.030520000000024083
524      51.50642000000005,-0.1272099999999341
527     51.615920000000074,-0.1767399999999384
528      51.48207000000008,0.07143000000002075
530      51.50645000000003,-0.2369099999999662
Name: post_code, Length: 310, dtype: object

Get the latitude

In [ ]:
lat_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[0])
lat_uk

0      51.492450000000076
1       51.51324000000005
6       51.51200000000006
7       51.51651000000004
9       51.41009000000008
              ...        
523    51.589770000000044
524     51.50642000000005
527    51.615920000000074
528     51.48207000000008
530     51.50645000000003
Name: post_code, Length: 310, dtype: object

Get the Longitude

In [ ]:
lng_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[1])
lng_uk

0       0.12127000000003818
1       -0.2674599999999714
6      -0.08057999999994081
7      -0.11967999999995982
9      -0.05682999999993399
               ...         
523    0.030520000000024083
524     -0.1272099999999341
527     -0.1767399999999384
528     0.07143000000002075
530     -0.2369099999999662
Name: post_code, Length: 310, dtype: object

We now have the geographical co-ordinates of the London Neighbourhoods.

We proceed with Merging our source data with the geographical co-ordinates to make our dataset ready for the next stage

In [ ]:
london_merged = pd.concat([df1,lat_uk.astype(float), lng_uk.astype(float)], axis=1)
london_merged.columns= ['borough','town','post_code','latitude','longitude']
london_merged

borough  ... longitude
0                Bexley, Greenwich   ...   0.12127
1    Ealing, Hammersmith and Fulham  ...  -0.26746
6                              City  ...  -0.08058
7                       Westminster  ...  -0.11968
9                           Bromley  ...  -0.05683
..                              ...  ...       ...
523                       Redbridge  ...   0.03052
524       Redbridge, Waltham Forest  ...  -0.12721
527                          Barnet  ...  -0.17674
528                       Greenwich  ...   0.07143
530          Hammersmith and Fulham  ...  -0.23691

[310 rows x 5 columns]

In [ ]:
london_merged.dtypes

borough       object
town          object
post_code     object
latitude     float64
longitude    float64
dtype: object

Getting the geocode for london to help visualize it on the map

In [ ]:
london = geocode(address='London, England, GBR')[0]
london_lng_coords = london['location']['x']
london_lat_coords = london['location']['y']
london_lng_coords

-0.1272099999999341

In [ ]:
london_lat_coords

51.50642000000005

### STEP 3: Visualize the Map of London

To help visualize the Map of London and the neighbourhoods in London, we make use of the folium package.

In [ ]:
# Creating the map of Toronto
map_London = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)
map_London

# adding markers to map
for latitude, longitude, borough, town in zip(london_merged['latitude'], london_merged['longitude'], london_merged['borough'], london_merged['town']):
    label = '{}, {}'.format(town, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_London)  
    
map_London

### Define Foursquare API credentials

To proceed with the next part, we need to define Foursquare API credentials

In [ ]:
CLIENT_ID = 'LDIJF4KI5VGMMA3NNDLFZWHR12TCMNTUL0TUC3QPZ3SJD040' 
CLIENT_SECRET = 'D5FECV3TLYRAYS5ROF0MGXFDLY3UBSQOZPOFHJRKTIUV542X'
VERSION = '20180605' # Foursquare API version

Defining a function to get the neraby venues in the neighbourhood

In [ ]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Getting the venues in London

In [ ]:
venues_in_London = getNearbyVenues(london_merged['borough'], london_merged['latitude'], london_merged['longitude'])

Bexley, Greenwich 
Ealing, Hammersmith and Fulham
City
Westminster
Bromley
Islington
Islington
Barnet
Enfield
Wandsworth
Southwark
City
Richmond upon Thames
Barnet
Islington
Wandsworth
Westminster
Bromley
Newham
Ealing
Westminster
Lewisham
Camden
Southwark
Tower Hamlets
Bexley
City
Lewisham
Greenwich
Tower Hamlets
Camden
Haringey
Tower Hamlets
Haringey
Barnet
Brent
Lambeth
Lewisham
Tower Hamlets
Kensington and ChelseaHammersmith and Fulham
Brent
Barnet
Barnet
Southwark
Tower Hamlets
Camden
Tower Hamlets
Waltham Forest
Newham
Islington
Richmond upon Thames
Lewisham
Camden
Westminster
Greenwich
Kensington and Chelsea
Barnet
Westminster
Lewisham
Waltham Forest
Hounslow, Ealing, Hammersmith and Fulham
Brent
Barnet
Lambeth, Wandsworth
Islington
Barnet
Merton
Barnet
Westminster
Barnet, Brent, Camden
Lewisham
Bexley
Haringey
Bromley
Tower Hamlets
Newham
Hackney
Dartford
Islington
Southwark
Lewisham
Brent
Southwark
Ealing
Kensington and Chelsea
Wandsworth
Southwark
Barnet
Newham
Richmond upon 

Sampling our data

In [ ]:
venues_in_London.head()

Neighbourhood  ...  Venue Category
0  Bexley, Greenwich   ...     Supermarket
1  Bexley, Greenwich   ...   Historic Site
2  Bexley, Greenwich   ...     Supermarket
3  Bexley, Greenwich   ...   Train Station
4  Bexley, Greenwich   ...     Coffee Shop

[5 rows x 5 columns]

In [ ]:
venues_in_London.shape

(6805, 5)

Wow, we have scraped together 6805 records for venues. This will definitely make the clustering interesting.

We need to now see how many Venue Categories are there for further processing

In [ ]:
venues_in_London.groupby('Venue Category').max()

Neighbourhood  ...               Venue
Venue Category                                                        ...                    
African Restaurant                                       Westminster  ...  Red Sea Restaurant
American Restaurant                                   Waltham Forest  ...        The Fat Bear
Antique Shop                                  Kensington and Chelsea  ...             Alice's
Arepa Restaurant                                       Tower Hamlets  ...          Arepa & Co
Argentinian Restaurant                                   Westminster  ...               Zoilo
...                                                              ...  ...                 ...
Wings Joint                                                   Camden  ...            Wingmans
Women's Store           Kensington and ChelseaHammersmith and Fulham  ...  Vivien of Holloway
Xinjiang Restaurant                                        Southwark  ...           Silk Road
Yoga Studio                                              Westminster  ...           yogahaven
Zoo Exhibit                                                   Camden  ...       Penguin Beach

[258 rows x 4 columns]

We can see 258 records, just goes to show how diverse and interesting the place is.

### One Hot Encoding 
We need to Encode our venue categories to get a better result for our clustering

In [ ]:
London_venue_cat = pd.get_dummies(venues_in_London[['Venue Category']], prefix="", prefix_sep="")
London_venue_cat

African Restaurant  American Restaurant  ...  Yoga Studio  Zoo Exhibit
0                      0                    0  ...            0            0
1                      0                    0  ...            0            0
2                      0                    0  ...            0            0
3                      0                    0  ...            0            0
4                      0                    0  ...            0            0
...                  ...                  ...  ...          ...          ...
6800                   0                    0  ...            0            0
6801                   0                    0  ...            0            0
6802                   0                    0  ...            0            0
6803                   0                    0  ...            0            0
6804                   0                    0  ...            0            0

[6805 rows x 258 columns]

Adding Neighbourhood into the mix.

In [ ]:
London_venue_cat['Neighbourhood'] = venues_in_London['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [London_venue_cat.columns[-1]] + list(London_venue_cat.columns[:-1])
London_venue_cat = London_venue_cat[fixed_columns]

London_venue_cat.head()

Neighbourhood  African Restaurant  ...  Yoga Studio  Zoo Exhibit
0  Bexley, Greenwich                    0  ...            0            0
1  Bexley, Greenwich                    0  ...            0            0
2  Bexley, Greenwich                    0  ...            0            0
3  Bexley, Greenwich                    0  ...            0            0
4  Bexley, Greenwich                    0  ...            0            0

[5 rows x 259 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood

In [ ]:
London_grouped = London_venue_cat.groupby('Neighbourhood').mean().reset_index()
London_grouped.head()

Neighbourhood  African Restaurant  ...  Yoga Studio  Zoo Exhibit
0                 Barnet                 0.0  ...          0.0          0.0
1  Barnet, Brent, Camden                 0.0  ...          0.0          0.0
2                 Bexley                 0.0  ...          0.0          0.0
3      Bexley, Greenwich                 0.0  ...          0.0          0.0
4     Bexley, Greenwich                  0.0  ...          0.0          0.0

[5 rows x 259 columns]


Let's make a function to get the top most common venue categories

In [ ]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]


There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods.

Creating a function to label the columns of the venue correctly

In [ ]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))


Getting the top venue categories in London

In [ ]:
# create a new dataframe for London
neighborhoods_venues_sorted_london = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_london['Neighbourhood'] = London_grouped['Neighbourhood']

for ind in np.arange(London_grouped.shape[0]):
    neighborhoods_venues_sorted_london.iloc[ind, 1:] = return_most_common_venues(London_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_london.head()

Neighbourhood  ...   10th Most Common Venue
0                 Barnet  ...              Pizza Place
1  Barnet, Brent, Camden  ...  Fruit & Vegetable Store
2                 Bexley  ...                 Bus Stop
3      Bexley, Greenwich  ...              Flower Shop
4     Bexley, Greenwich   ...            Fishing Store

[5 rows x 11 columns]

### STEP : Model Building

Let's cluster the city of london to roughly 5 to make it easier to analyze. 

We use the K Means clustering technique to do so.

In [ ]:
# set number of clusters
k_num_clusters = 5

London_grouped_clustering = London_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans_london = KMeans(n_clusters=k_num_clusters, random_state=0).fit(London_grouped_clustering)
kmeans_london

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=5, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=0, tol=0.0001, verbose=0)

In [ ]:
kmeans_london.labels_

array([1, 3, 3, 3, ..., 1, 1, 1, 1], dtype=int32)

So our model has labeled the city

In [ ]:
neighborhoods_venues_sorted_london.insert(0, 'Cluster Labels', kmeans_london.labels_)

Join London_merged with our neighbourhood venues sorted to add latitude & longitude for each of the neighborhood to prepare it for plotting

In [ ]:
london_data = london_merged

london_data = london_data.join(neighborhoods_venues_sorted_london.set_index('Neighbourhood'), on='borough')

london_data.head()

borough  ... 10th Most Common Venue
0              Bexley, Greenwich   ...          Fishing Store
1  Ealing, Hammersmith and Fulham  ...            Flea Market
6                            City  ...    Japanese Restaurant
7                     Westminster  ...      Indian Restaurant
9                         Bromley  ...              Gastropub

[5 rows x 16 columns]


Drop all the NaN values to prevent data skew

In [ ]:
london_data_nonan = london_data.dropna(subset=['Cluster Labels'])

Let's plot the clusters

In [ ]:
map_clusters_london = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(london_data_nonan['latitude'], london_data_nonan['longitude'], london_data_nonan['borough'], london_data_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters_london)
        
map_clusters_london

### Model Evaluation

Cluster 1

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 0, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels  ... 9th Most Common Venue 10th Most Common Venue
260  LONDON               0  ...                  Park           Climbing Gym

[1 rows x 12 columns]

Cluster 2

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 1, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  ...  10th Most Common Venue
6                    LONDON  ...     Japanese Restaurant
7                    LONDON  ...       Indian Restaurant
10                   LONDON  ...    Gym / Fitness Center
12                   LONDON  ...    Gym / Fitness Center
14           BARNET, LONDON  ...             Pizza Place
..                      ...  ...                     ...
523                  LONDON  ...             Coffee Shop
524  LONDON, WOODFORD GREEN  ...       Outdoor Sculpture
527                  LONDON  ...             Pizza Place
528                  LONDON  ...                    Park
530                  LONDON  ...             Yoga Studio

[294 rows x 12 columns]

Cluster 3

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 2, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  ...  10th Most Common Venue
379  HARROW, STANMOREEDGWARE, LONDON  ...       Fish & Chips Shop

[1 rows x 12 columns]

Cluster 4

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 3, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  ...   10th Most Common Venue
0                 LONDON  ...            Fishing Store
9                 LONDON  ...                Gastropub
29     BECKENHAM, LONDON  ...                Gastropub
45   BEXLEYHEATH, LONDON  ...                 Bus Stop
121               LONDON  ...  Fruit & Vegetable Store
124               LONDON  ...                 Bus Stop
127               LONDON  ...                Gastropub
168      LONDON, WELLING  ...              Flower Shop
292       LONDON, SIDCUP  ...                 Bus Stop
317               LONDON  ...                Gastropub
361               LONDON  ...                Gastropub
459        LONDON, ERITH  ...              Flower Shop
507               LONDON  ...                 Bus Stop

[13 rows x 12 columns]

Cluster 5

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 4, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels  ... 9th Most Common Venue 10th Most Common Venue
1  LONDON               4  ...         Fishing Store            Flea Market

[1 rows x 12 columns]

The neighbourhoods of London are very mulitcultural. They have a lot of different cusines including Indian, Italian, Turkish and Chinese. London seems to take a step further in this direction by having a lot of Restaurants, bars, juice bars, coffee shops, Fish and Chips shop and Breakfast spots. It has a lot of shopping options too with that of the Flea markets, flower shops, fish markets, Fishing stores, clothing stores. The main modes of transport seem to be Buses and trains. For leisure, the neighbourhoods are set up to have lots of parks, golf courses, zoo, gyms and Historic sites. 
<br>
<br>
Overall, the city of London offers a multicural, diverse and certainly an entertaining experience.